sm_ipt: Add Shared Memory Inter-Processor Transport (#524)
nrfconnect/sdk-nrfxlib: sm_ipt/doc/usage.rst

	int send_nocopy(size_t len)
	{
		uint8_t *buffer;
		size_t i;

		sm_ipt_alloc_tx_buf(&sm_ipt_ctx, &buffer, len);

		/* Use buffer as any other allocated buffer */
		some_data_preparation_function(buffer);

		return sm_ipt_send(&sm_ipt_ctx, buffer, len);
	}

The above code uses the :c:func:`sm_ipt_alloc_tx_buf` function to allocate the transmit buffer in the shared memory.
The buffer is prepared by the :c:func:`some_data_preparation_function` and sent to another core through the :c:func:`sm_ipt_send` function.

Data send abort
---------------

It is possible to free the transmit buffer instead of sending it.

.. code-block:: c

	int send_if(size_t len)
	{
		uint8_t *buffer;
		bool preparation_success;

		sm_ipt_alloc_tx_buf(&sm_ipt_ctx, &buffer, len);

		/* Prepare some data */
		preparation_success = some_data_preparation_function(buffer);

		if (preparation_success) {
			return sm_ipt_send(&sm_ipt_ctx, buffer, len);
		} else {
			sm_ipt_free_tx_buf(&sm_ipt_ctx, buffer);

			return ERROR_ABORT;
		}
	}

The above code uses the :c:func:`sm_ipt_alloc_tx_buf` function to allocate the transmit buffer in the shared memory.
The buffer is sent only if ``preparation_success`` is true.

> If ``preparation_success`` is false the buffer is free using the :c:func:`sm_ipt_free_tx_buf` and the send operation is aborted.

Reviewer: ```suggestion
If ``preparation_success`` is false, the buffer is freed using the :c:func:`sm_ipt_free_tx_buf` function and the send operation is aborted.
```

		data_received = true;
	}

The above code uses the :c:func:`write_bytes_to_pipe` function to copy the shared memory buffer into the local pipe.
The shared memory buffer is freed using the :c:func:`sm_ipt_free_rx_buf` function.
The data is processed locally using a copy of the shared memory buffer.

Data receive no-copy
--------------------

When the data is received, it can be processed using the no-copy mechanism.
The pointer to the shared memory buffer must be saved for later use.
The shared memory rx buffer must be freed and data processing takes place on the shared memory buffer.

.. note::
   This approach increases the risk of out-of-memory errors if the buffer is not freed right after receiving data.

.. code-block:: c

	void sm_ipt_receive_handler(const uint8_t *packet, size_t len)
	{
		/* Put the buffer pointer to some FIFO */
		put_buffer_ptr_to_fifo(packet, len);
	}

	void data_processing_thread(void)
	{
		uint8_t *packet;
		size_t len;

		/* Wait for incoming data */
		get_buffer_ptr_from_fifo(&packet, &len);

		/*Process data */
		some_data_processing_function(packet, len);

		/* Free receive buffer */
		sm_ipt_free_rx_buf(&sm_ipt_ctx, packet);
	}

> The above code puts the pointer to the shared memory buffer to the local fifo using :c:func:`put_buffer_ptr_to_fifo` function.

Reviewer: ```suggestion
The above code sets the pointer to the shared memory buffer to the local fifo using the :c:func:`put_buffer_ptr_to_fifo` function.
```

		memcpy(buffer, data, len);

		return sm_ipt_send(&sm_ipt_ctx, buffer, len);
	}

The above code uses the :c:func:`sm_ipt_alloc_tx_buf` function to allocate the transmit buffer in the shared memory.
The data is copied into the said buffer and sent to another core through the :c:func:`sm_ipt_send` function.

Data send no-copy
-----------------

It is possible to operate directly on the transmit buffer to avoid copying from the local buffer.

.. code-block:: c

	int send_nocopy(size_t len)
	{
		uint8_t *buffer;
		size_t i;

		sm_ipt_alloc_tx_buf(&sm_ipt_ctx, &buffer, len);

		/* Use buffer as any other allocated buffer */
		some_data_preparation_function(buffer);

		return sm_ipt_send(&sm_ipt_ctx, buffer, len);
	}

The above code uses the :c:func:`sm_ipt_alloc_tx_buf` function to allocate the transmit buffer in the shared memory.
The buffer is prepared by the :c:func:`some_data_preparation_function` and sent to another core through the :c:func:`sm_ipt_send` function.

Data send abort
---------------

It is possible to free the transmit buffer instead of sending it.

.. code-block:: c

	int send_if(size_t len)
	{
		uint8_t *buffer;
		bool preparation_success;

> 		size_t i;

Reviewer 2: `i` is unused.

.. _sm_ipt_usage:

> Usage and Implementation

Reviewer: ```suggestion
Usage and implementation
```
We don't use title case

########################

.. contents::
   :local:
   :depth: 2

The following picture gives an overview of the SM IPT architecture:

.. figure:: img/architecture.svg
   :alt: SM IPT architecture layers
   :align: center

   SM IPT architecture layers

Usage
=====

.. note::

>    Following examples are purely conceptual and simplified to show library functions. They should be rewritten to suit particular need and use OS modules like FIFO.

Reviewer: ```suggestion
   Following examples are purely conceptual and simplified to demonstrate the library functions.
   They should be rewritten to suit particular need and use OS modules like FIFO.
```
Every sentence must start from a new line. Everywhere.

Initialization
--------------

> The initialization function initializes context data and retrieves shared memory parameters from ``sm_ipt_os_init`` function.

Reviewer: ```suggestion
The initialization function initializes the context data and retrieves the shared memory parameters from the :c:func:`sm_ipt_os_init` function.
```

.. code-block:: c

	struct sm_ipt_ctx sm_ipt_ctx;

	void sm_ipt_receive_handler(const uint8_t *packet, size_t len)
	{
		/* Process data */
		some_data_processing_function();
	}

	int initialize(void)
	{
		return sm_ipt_init(&sm_ipt_ctx, sm_ipt_receive_handler);
	}

> The above code uses the ``sm_ipt_init`` function to initialize ``struct sm_ipt_ctx`` SM IPT context data structure and assign the receive handler.

Reviewer: ```suggestion
The above code uses the :c:func:`sm_ipt_init` function to initialize the :c:struct:`struct sm_ipt_ctx` SM IPT context data structure and to assign the receive handler.
```

Data send full-copy
-------------------

> The data is prepared beforehand, copied into allocated buffer and send to another core.

Reviewer: ```suggestion
The data is prepared beforehand, copied into the allocated buffer and sent to another core.
```

.. code-block:: c

	int send(const void *data, size_t len)
	{
		uint8_t *buffer;

		sm_ipt_alloc_tx_buf(&sm_ipt_ctx, &buffer, len);

		memcpy(buffer, data, len);

		return sm_ipt_send(&sm_ipt_ctx, buffer, len);
	}

The above code uses the ``sm_ipt_alloc_tx_buf`` function to allocate transmit buffer in shared memory. Data is copied into said buffer and sent to another core via ``sm_ipt_send`` function.

Data send no-copy

> -------------------

Reviewer: Match the underlining with the title text

> The data is prepared after buffer allocation send to another core when data preparation (and usage) is finished.

Reviewer: This is very unclear. Please clarify.

.. code-block:: c

	int send_nocopy(size_t len)
	{
		uint8_t *buffer;
		size_t i;

		sm_ipt_alloc_tx_buf(&sm_ipt_ctx, &buffer, len);

		/* Use buffer as any other allocated buffer */
		some_data_preparation_function(buffer);

		return sm_ipt_send(&sm_ipt_ctx, buffer, len);
	}

> The above code uses the ``sm_ipt_alloc_tx_buf`` function to allocate transmit buffer in shared memory. Buffer is prepared by ``some_data_preparation_function`` and sent to another core via ``sm_ipt_send`` function.

Reviewer: ```suggestion
The above code uses the :c:func:`sm_ipt_alloc_tx_buf` function to allocate the transmit buffer in the shared memory.
The buffer is prepared by the :c:func:`some_data_preparation_function` and sent to another core through the :c:func:`sm_ipt_send` function.
```

Data send abort

> -------------------

Reviewer: Match the underlining with the title text

> There is possibility to free transmit buffer instead of sending it if the need arises.

Reviewer: ```suggestion
It is possible to free the transmit buffer instead of sending it.
```

.. code-block:: c

	int send_if(size_t len, bool send_choice)
	{
		uint8_t *buffer;
		size_t i;

		sm_ipt_alloc_tx_buf(&sm_ipt_ctx, &buffer, len);

		if (send_choice){
			/* Prepare some data */
			some_data_preparation_function(buffer);

			return sm_ipt_send(&sm_ipt_ctx, buffer, len);
		}
		else{
			sm_ipt_free_tx_buf(&sm_ipt_ctx, buffer);

			return ERROR_ABORT;
		}
	}

> The above code uses the ``sm_ipt_alloc_tx_buf`` function to allocate transmit buffer in shared memory. Buffer is send only if send_choice is true. If send_choice is false the buffer is free using ``sm_ipt_free_tx_buf`` and send operation is aborted.

Reviewer: ```suggestion
The above code uses the :c:func:`sm_ipt_alloc_tx_buf` function to allocate the transmit buffer in the shared memory.
The buffer is sent only if ``send_choice`` is true. If ``send_choice`` is false the buffer is free using the :c:func:`sm_ipt_free_tx_buf` and the send operation is aborted.
```

Data receive full-copy

> -------------------

Reviewer: The length of the underlining must match the number of characters in the title text

> When data is received it can be processed using full-copy mechanism. Buffer has to be copied to local memory. Shared memory rx buffer has to be freed and data processing is done on local copy of buffer.

Reviewer: ```suggestion
When data is received, it can be processed using the full-copy mechanism.
The buffer must be copied to the local memory.
The shared memory rx buffer must be freed and data processing takes place on the local copy of the buffer.
```

.. code-block:: c

	volatile uint8_t receive_buffer[MAX_PACKET_SIZE];
	volatile bool data_received;

	void sm_ipt_receive_handler(const uint8_t *packet, size_t len)
	{
		/* Copy whole message */
		memcpy(receive_buffer, packet, len);

		/* Free receive buffer */
		sm_ipt_free_rx_buf(&sm_ipt_ctx, packet);

		/* Signal message received */
		data_received = true;
	}

	void data_processing_thread(void)
	{
		/* Wait for incoming data */
		while(!data_received);

		/*Process data */
		some_data_processing_function(receive_buffer);
	}

> The above code uses the ``memcpy`` function to copy shared memory buffer into local buffer. Next shared memory buffer is freed using ``sm_ipt_free_rx_buf``. Lastly the data is processed locally using a copy of buffer.

Reviewer: ```suggestion
The above code uses the :c:func:`memcpy` function to copy the shared memory buffer into the local buffer.
The next shared memory buffer is freed using the :c:func:`sm_ipt_free_rx_buf` function
 The data is processed locally using a copy of the buffer.
```

Data receive no-copy

> -------------------

Reviewer: Match the underlining with the title text

> When data is received it can be processed using no-copy mechanism. The pointer to shared memory buffer has to be saved in order to use it later. Shared memory rx buffer has to be freed and data processing is done on shared memory buffer.

Reviewer: ```suggestion
When the data is received, it can be processed using the no-copy mechanism.
The pointer to the shared memory buffer must be saved for later use.
The shared memory rx buffer must be freed and data processing takes place on the shared memory buffer.
```

.. note::

>    This approach increases risk of out of memory errors due to not freeing buffer right after receiving data.

Reviewer: ```suggestion
   This approach increases the risk of out-of-memory errors if the buffer is not freed right after receiving data.
```

.. code-block:: c

	volatile void *receive_buffer;
	volatile bool data_received;

	void sm_ipt_receive_handler(const uint8_t *packet, size_t len)
	{
		/* Save buffer pointer */
		receive_buffer = (void *) packet;

		/* Signal message received */
		data_received = true;
	}

	void data_processing_thread(void)
	{
		/* Wait for incoming data */
		while(!data_received);

		/*Process data */
		some_data_processing_function(receive_buffer);

		/* Free receive buffer */
		sm_ipt_free_rx_buf(&sm_ipt_ctx, packet);
	}

> The above code copies pointer to shared memory buffer. Next data is processed using original buffer in share memory. Lastly shared memory buffer is freed using ``sm_ipt_free_rx_buf``.

Reviewer: ```suggestion
The above code copies the pointer to the shared memory buffer.
Next data is processed using the original buffer in the shared memory.
The shared memory buffer is freed using the :c:func:`sm_ipt_free_rx_buf` function.
```

Lower layers
============

The lower layers of SM IPT are OS-dependent.
They are responsible for memory management, synchronization and logging support.

.. note::

>    Detailed knowledge about how the lower layers are implemented is not required when using the SM IPT API.

Reviewer: ```suggestion
   Details about the implementation of the lower layers are not required when using the SM IPT API.
```

>    However, this knowledge is required to port to a different operating system.

Reviewer: ```suggestion
   However, this knowledge is needed to port to a different operating system.
```

Operating system abstraction
----------------------------

> The operating system abstraction provides functionality for SM IPT that depends on the operating system.

Reviewer: ```suggestion
The operating system abstraction provides the SM IPT functionalities that depend on the operating system.
```

The initialization function initializes the context data and retrieves the shared memory parameters from the :c:func:`sm_ipt_os_init` function.

.. code-block:: c

	struct sm_ipt_ctx sm_ipt_ctx;

	void sm_ipt_receive_handler(const uint8_t *packet, size_t len)
	{
		/* Process data */
		some_data_processing_function();
	}

	int initialize(void)
	{
		return sm_ipt_init(&sm_ipt_ctx, sm_ipt_receive_handler);
	}

The above code uses the :c:func:`sm_ipt_init` function to initialize the :c:struct:`sm_ipt_ctx` SM IPT context data structure and to assign the receive handler.

Data send full-copy
-------------------

The data is prepared beforehand, copied into the allocated buffer and sent to another core.

.. code-block:: c

	int send(const void *data, size_t len)
	{
		uint8_t *buffer;

		sm_ipt_alloc_tx_buf(&sm_ipt_ctx, &buffer, len);

		memcpy(buffer, data, len);

		return sm_ipt_send(&sm_ipt_ctx, buffer, len);
	}

The above code uses the :c:func:`sm_ipt_alloc_tx_buf` function to allocate the transmit buffer in the shared memory.
The data is copied into the said buffer and sent to another core through the :c:func:`sm_ipt_send` function.

Data send no-copy
-----------------

> It is possible to directly operate on the transmit buffer in order to avoid copying from the local buffer.

Reviewer: ```suggestion
It is possible to operate directly on the transmit buffer to avoid copying from the local buffer.
```

Data send abort
---------------

It is possible to free the transmit buffer instead of sending it.

.. code-block:: c

	int send_if(size_t len, bool send_choice)
	{
		uint8_t *buffer;
		size_t i;

		sm_ipt_alloc_tx_buf(&sm_ipt_ctx, &buffer, len);

		if (send_choice){
			/* Prepare some data */
			some_data_preparation_function(buffer);

			return sm_ipt_send(&sm_ipt_ctx, buffer, len);
		}
		else{
			sm_ipt_free_tx_buf(&sm_ipt_ctx, buffer);

			return ERROR_ABORT;
		}
	}

The above code uses the :c:func:`sm_ipt_alloc_tx_buf` function to allocate the transmit buffer in the shared memory.
The buffer is sent only if ``send_choice`` is true. If ``send_choice`` is false the buffer is free using the :c:func:`sm_ipt_free_tx_buf` and the send operation is aborted.

Data receive full-copy
----------------------

When data is received, it can be processed using the full-copy mechanism.
The buffer must be copied to the local memory.
The shared memory rx buffer must be freed and data processing takes place on the local copy of the buffer.

.. code-block:: c

	volatile uint8_t receive_buffer[MAX_PACKET_SIZE];

> 	volatile bool data_received;

Reviewer 2: I am not sure that using volatiles int this sample is a good idea. Maybe use some generic functions names, e.g. 

```c
/* Write content of the buffer to some pipe. */
write_bytes_to_pipe(packet, len);

/* Free receive buffer */
sm_ipt_free_rx_buf(&sm_ipt_ctx, packet);
```

And then, `data_processing_thread` is not needed.

	void sm_ipt_receive_handler(const uint8_t *packet, size_t len)
	{
		/* Copy whole message */
		memcpy(receive_buffer, packet, len);

		/* Free receive buffer */
		sm_ipt_free_rx_buf(&sm_ipt_ctx, packet);

		/* Signal message received */
		data_received = true;
	}

	void data_processing_thread(void)
	{
		/* Wait for incoming data */
		while(!data_received);

		/*Process data */
		some_data_processing_function(receive_buffer);
	}

The above code uses the :c:func:`memcpy` function to copy the shared memory buffer into the local buffer.
The next shared memory buffer is freed using the :c:func:`sm_ipt_free_rx_buf` function
The data is processed locally using a copy of the buffer.

Data receive no-copy
--------------------

When the data is received, it can be processed using the no-copy mechanism.
The pointer to the shared memory buffer must be saved for later use.
The shared memory rx buffer must be freed and data processing takes place on the shared memory buffer.

.. note::
   This approach increases the risk of out-of-memory errors if the buffer is not freed right after receiving data.

.. code-block:: c

	volatile void *receive_buffer;

> 	volatile bool data_received;

Reviewer 2: The same comment as above. Maybe something like that:

```c
/* Put the buffer pointer to some FIFO */
put_buffer_ptr_to_fifo(packet, len);
```

in `data_processing_thread`:

```c
/* Wait for incoming data */
get_buffer_ptr_from_fifo(&packet, &len);

/*Process data */
some_data_processing_function(receive_buffer);

/* Free receive buffer */
sm_ipt_free_rx_buf(&sm_ipt_ctx, packet);

```
